### PR TITLE
Add Kubernetes IngressClass support

### LINKED ISSUE #1397: Kubernetes Ingress Class Support
### Background

[Kubernetes 1.18 introduced](https://kubernetes.io/blog/2020/04/02/improvements-to-the-ingress-api-in-kubernetes-1.18/) a standard mechanism for assigning individual Ingress resources to specific Ingress Controllers, using the `ingressClassName` field on Ingresses and the *IngressClass* resource. Up to that point, Ingress Controllers used a custom annotation `kubernetes.io/ingress.class` on Ingress resources but in subtly different ways.

Kubernetes 1.22 (released Oct 2021) removed the beta Ingress resource API which means that all Ingress resources support the `ingressClassName` field.

### Why is this important?

Support for `IngressClass` means that groups of Ingress resources can be explicitly allocated to specific instances of an Ingress Controller, thereby segregating your traffic.

For example, you may run two Ingress Controllers in a cluster: one for external access, the other for internal access. You could set the `ingressClassName` field for your public APIs to `yarp-public` , and your monitoring Ingresses to `yarp-internal`.

Another example would be to run multiple Ingress Controller deployments/daemonsets to help distribute the load more effectively.

```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: the-ingress
  namespace: default
spec:
  ingressClassName: yarp
  rules:
    - http:
        paths:
          - path: /
            pathType: ImplementationSpecific
            backend:
              service:
                name: the-service
                port: 
                  name: http
```

### PR DESCRIPTION
**Pull request by illay1994**
Fixes #1397

- Added informer for IngressClass resources
- Updated RBAC in sample manifest to allow list/read of IngressClass resources
- Delayed start of Ingress resource watching until IngressClass, Service, and Endpoints are synchronized
- Modified ingress cache to only store Ingresses that match the IngressClass rules
- Bound the existing ConfigMap to Application Settings and included sensible default for Ingress Class name (this allows the value to be changed when multiple, independent instances of the Yarp Ingress Controller are deployed into the same cluster)

### CODE REVIEW

**Comment by illay1994** [comment]
#1398 Fixed PR Comments

**Comment by MihaZupan** [comment]
@dpbevin did you want to continue your PR (#1398) or are you okay with building on that work here?

**Comment by dpbevin** [comment]
@MihaZupan I'm ok with this being merged in instead of #1398.

**Review by MihaZupan — APPROVED** [review]
With limited understanding of the details here, this LGTM.

@dpbevin if you get a chance can you give this one a once-over (I realize most of the changes are from your original PR)?

File: src/Kubernetes.Controller/Caching/IngressCache.cs [hunk 1]
```diff
@@ -18,16 +21,79 @@ namespace Yarp.Kubernetes.Controller.Caching;
...
+        if (ingressClass is null)
+        {
+            throw new ArgumentNullException(nameof(ingressClass));
+        }
+
+        if (!string.Equals(_options.ControllerClass, ingressClass.Spec.Controller, StringComparison.OrdinalIgnoreCase))
+        {
+#pragma warning disable CA1303 // Do not pass literals as localized parameters
+            _logger.LogInformation(
+                "Ignoring {IngressClassNamespace}/{IngressClassName} as the spec.controller is not the same of this ingress",
```
[the remark this answers is not in the document]
Comment (illay1994) [reply src/Kubernetes.Controller/Caching/IngressCache.cs]: Fixed

File: src/Kubernetes.Controller/Caching/IngressCache.cs [hunk 2]
```diff
@@ -77,6 +143,16 @@ public IEnumerable<IngressData> GetIngresses()
         return ingresses;
     }
 
+    private bool IsYarpIngress(V1IngressSpec spec)
+    {
+        if (spec.IngressClassName != null)
+        {
+            return _ingressClassData.ContainsKey(spec.IngressClassName);
```
[the remark this answers is not in the document]
Comment (illay1994) [reply src/Kubernetes.Controller/Caching/IngressCache.cs]: Thanks for suggestions. Fixed

File: src/Kubernetes.Controller/Caching/IngressClassData.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            if (ingressClass is null)
+            {
+                throw new ArgumentNullException(nameof(ingressClass));
+            }
+
+            IngressClass = ingressClass;
+            IsDefault = GetDefaultAnnotation(ingressClass);
+        }
+
+        public V1IngressClass IngressClass { get; set; }
```
[the remark this answers is not in the document]
Comment (illay1994) [reply src/Kubernetes.Controller/Caching/IngressClassData.cs]: Thanks for suggestions. Fixed

**Comment by dpbevin** [comment]
@MihaZupan this still looks good to me. Once merged, I'll check that all of the recent changes are working together nicely and make a start on the next round of changes.

**Pull request merged**